Switch to pnpm causes runtime dependency error: Cannot find module '@directus/app/package.json'

### Describe the Bug

The switch to pnpm seems to have changed the way that workspace dependencies are resolved.

When starting directus by running api/cli.js start, the error: Cannot find module '@directus/app/package.json' is shown and causes a crash.

It looks like workspace:* dependencies are not being resolved correctly in resolve-packages.ts - the call to require.resolve is using the node module resolution algorithm relative to /shared/util rather than relative to the main module or peer dependencies.

Current temporary workaround is to add @directus/app as a dependency to the shared package, this will allow directus to start, but isn't going to work for package extension resolution.



### To Reproduce

Do a fresh git clone of main branch, pnpm install, pnpm -r build, and try to run via api/cli.js start

### Errors Shown

Cannot find module '@directus/app/package.json'

### What version of Directus are you using?

9.14.5

### What version of Node.js are you using?

16

### What database are you using?

Postgres 14

### What browser are you using?

Chrome

### How are you deploying Directus?

local

--- a/api/src/app.ts
+++ b/api/src/app.ts
@@ -179,7 +179,7 @@ export default async function createApp(): Promise<express.Application> {
 	});
 
 	if (env.SERVE_APP) {
-		const adminPath = require.resolve('@directus/app');
+		const adminPath = require.resolve('@directus/app', require.main ? { paths: [require.main.filename] } : undefined);
 		const adminUrl = new Url(env.PUBLIC_URL).addPath('admin');
 
 		// Set the App's base path according to the APIs public URL

--- a/api/src/extensions.ts
+++ b/api/src/extensions.ts
@@ -310,7 +310,9 @@ class ExtensionManager {
 	}
 
 	private async getSharedDepsMapping(deps: string[]) {
-		const appDir = await fse.readdir(path.join(resolvePackage('@directus/app'), 'dist', 'assets'));
+		const appDir = await fse.readdir(
+			path.join(resolvePackage('@directus/app', require.main?.filename), 'dist', 'assets')
+		);
 
 		const depsMapping: Record<string, string> = {};
 		for (const dep of deps) {

